HTTP/3 graceful shutdown

hfujita opened the pull request:
This is an initial attempt to implement HTTP/3 graceful shutdown (#2394).

# What is supported in this PR

* send `GOAWAY` frame upon graceful shutdown request (i.e. `SIGTERM`)
    * resend GOAWAY after 1 second from the initial GOAWAY
* `http3-graceful-shutdown-timeout` configuration support, as in HTTP/2
* minimal GOAWAY frame validation (check frame length, etc.) on client side when receiving GOAWAY frame
* simple test case that ensures a client actually receives GOAWAY when a server is shutting down

# What is NOT supported in this PR

* Server side: rejecting further requests after sending GOAWAY
* Client side: closing connection after receiving GOAWAY (and thus suppressing further requests)

# Discussion

The GOAWAY frame encoder (`frame.c: h2o_http3_encode_goaway_frame`) now follows the same way as  `h2o_http3_encode_priority_update_frame`. There are a few alternatives we can take.

* Pass `h2o_buffer_t *` to the encoder function instead of `uint8_t *` (as in `h2o_http2_encode_goaway_frame`). This will encapsulate the buffer management (allocation) inside the encoder function, which makes the caller-side code simpler.
* Use `h2o_http3_encode_frame` macro -- I haven't fully figured out if this is actually a viable option though.

Let me know if you have a preference on one way over others.

# Known issue

* The test case fails in my local macOS environment (Mojave). It appears that `dtrace -p ${client_pid}` is attached to the server process for some reason, thus emitting incorrect strings. Will see if it also happens in CI.

kazuho commented on lib/http3/frame.c:
```diff
@@ -59,3 +59,37 @@ int h2o_http3_decode_priority_update_frame(h2o_http3_priority_update_frame_t *fr
+
+    /* quicly_decodev below will reject len == 0 case */
+    if (len > PTLS_ENCODE_QUICINT_CAPACITY)
+        goto Fail;
```
I think this error check is insufficient. Consider the case where the given payload was `0100` (len = 2 bytes). This should be considered invalid, as the second byte is a junk.

I think the correct thing to do in this function is to assert `src == end` after returning successfully from `quicly_decodev`.

hfujita replied:
👍 I totally missed that case. Thanks!

kazuho commented on lib/http3/server.c:
```diff
@@ -1473,6 +1476,8 @@ h2o_http3_conn_t *h2o_http3_server_accept(h2o_http3_server_ctx_t *ctx, quicly_ad
     conn->scheduler.uni.active = 0;
     conn->scheduler.uni.conn_blocked = 0;
+    memset(&conn->_conns, 0, sizeof(conn->_conns));
+    h2o_linklist_insert(&ctx->accept_ctx->ctx->http3._conns, &conn->_conns);
```
Please use `h2o_linklist_init_anchor`.

hfujita replied:
Do you mean this?
```
h2o_linklist_init_anchor(&conn->_conns);
```
I'm afraid it would cause an assertion failure `!h2o_linklist_is_linked(node)`  in `h2o_linklist_insert`.

kazuho replied:
Please use `h2o_linklist_is_empty` for checking if an anchor refers to something.

kazuho replied:
Ah my bad, `conn->_conns` is a node element, not an anchor. Yes, what you have now is fine.

hfujita replied:
Got it. Thank you for checking.

kazuho commented on lib/http3/server.c:
```diff
@@ -1500,8 +1505,71 @@ h2o_http3_conn_t *h2o_http3_server_accept(h2o_http3_server_ctx_t *ctx, quicly_ad
+    h2o_linklist_t *node, *next;
+
+    /* We've sent two GOAWAY frames, close the remaining connections */
+    for (node = ctx->http3._conns.next; node != &ctx->http3._conns; node = node->next) {
```
I think that the last expression should be `node = next`?

hfujita replied:
👍 Good catch, thanks! I thought I simply copied this logic from H2 but for some reason I introduced a mutation.

kazuho commented on lib/http3/server.c:
```diff
@@ -1500,8 +1505,71 @@ h2o_http3_conn_t *h2o_http3_server_accept(h2o_http3_server_ctx_t *ctx, quicly_ad
+    /* After waiting a second, we still had active connections. If configured, wait one
+     * final timeout before closing the connections */
+    if (do_close_stragglers && ctx->globalconf->http3.graceful_shutdown_timeout) {
+        h2o_timer_unlink(&ctx->http3._graceful_shutdown_timeout);
```
This line is unnecessary. By the timer invokes its callback, it is guaranteed to have been unlinked.

hfujita replied:
I see. Actually the same code exists in http2. Can we remove it from http2 as well?

hfujita replied:
Removed from http2 as well. We can easily revert it if I was not supposed to.

kazuho replied:
Ah thanks. I'll fix the http2 side separately.

kazuho commented on lib/http3/server.c:
```diff
@@ -1500,8 +1505,71 @@ h2o_http3_conn_t *h2o_http3_server_accept(h2o_http3_server_ctx_t *ctx, quicly_ad
+
+    /* After waiting a second, we still had active connections. If configured, wait one
+     * final timeout before closing the connections */
+    if (do_close_stragglers && ctx->globalconf->http3.graceful_shutdown_timeout) {
```
Maybe add `!= 0` to the second condition, as the value is `uint64_t`?

hfujita replied:
👍

kazuho commented on t/90http3-graceful-shutdown.t:
```diff
@@ -0,0 +1,131 @@
+
+if ($client_pid == 0) {
+	close STDERR;
+	open STDERR, ">", "/dev/null";
```
I think we might want to keep STDERR as-is, so that we could figure out when something goes wrong?

hfujita replied:
Done.

kazuho commented on t/90http3-graceful-shutdown.t:
```diff
@@ -0,0 +1,131 @@
+	} else {
+		exec(
+			qw(unbuffer dtrace -p), $client_pid, "-n", <<'EOT',
+:h2o::h3_frame_receive {
```
The 2nd element denotes the process name. In this case, it should be `h2o-httpclient`.

hfujita replied:
💯 Finally it all made sense! Thanks!

hfujita commented on t/90dtrace.t:
```diff
@@ -106,27 +91,7 @@ EOT
-    die "bpftrace failed to start\n"
-        if waitpid($tracer_pid, WNOHANG) == $tracer_pid;
-}
+$read_trace = get_tracer($tracer_pid, "$tempdir/trace.out");
```
Done.

hfujita commented:
@kazuho Thank you for the review. I think I've addressed your comments above, except for the `h2o_linklist_init_anchor` one. Also the test in macOS has been fixed 👍 Please take another look.

kazuho reviewed (approved):
Thank you for the changes. To be merged.
